File: striker_poc/Moles/src/BaseMole.cpp
//------------------------------------
// ImperfectlyCoded © 2018
//
// BaseMole.cpp
//------------------------------------
 #include <iostream>
 #include "../include/BaseMole.h"
 #include "../Holes/Hole.h"
 
//-------------------------------
// Constructor(s) / Destructor
//-------------------------------
BaseMole::BaseMole()
: m_health(0)
, m_stamina(0)
, m_mound(0)
, m_state(0)
, m_type("BaseMole")
{
    std::cout << m_type.c_str() << " created!\n";
}

BaseMole::BaseMole(int health, int stamina, const Mound* mound)
: m_health(health)
, m_stamina(stamina)
, m_mound(mound)
, m_type("BaseMole")
{}

BaseMole::~BaseMole()
{
    m_mound = 0;
    m_state = 0;
}

//---------------------------------------------
// EnterState()
//
// state - the next state to enter
//---------------------------------------------
void BaseMole::EnterState(BaseState* state)
{
}

//---------------------------------------------
// ExitState()
//---------------------------------------------
void BaseMole::ExitState()
{
}
    
//-----------------------------------
// MoveInMound
//
// Moves mole into the mound.
//-----------------------------------
void BaseMole::MoveInMound()
{
    std::cout << m_type.c_str() << " moves into mound\n";
}

//-----------------------------------
// MoveOutMound
//
// Moves mole out of the mound.
//-----------------------------------
void BaseMole::MoveOutMound()
{
    std::cout << m_type.c_str() << " moves out of mound\n";
}

//-------------------------------------
// MoveToDifferentMound
//
// Moves mole to a different mound.
//
// mound - the new mound
//-------------------------------------
void BaseMole::MoveToDifferentMound(const Mound* mound)
{
    std::cout << m_type.c_str() << " moving to a different mound\n";
}

void BaseMole::ThrowProjectile()
{
    std::cout << m_type.c_str() << " throws a projectile\n";
}
    
void BaseMole::CounterAttack()
{
    std::cout << m_type.c_str() << " counter attacks\n";
}

void BaseMole::ChargeAttack()
{
    std::cout << m_type.c_str() << " charges to attack\n";
}

void BaseMole::EvadeAttack()
{
    std::cout << m_type.c_str() << " evades attack\n";
}

/***********************************
 * AdjustStamina
 *
 * Increments/Decrements stamina by 
 * change amount.
 *
 * changeAmt - the amount to increment
 *           or decrement stamina.
 ***********************************/
void BaseMole::AdjustStamina(int changeAmt)
{
    //m_health += changeAmt;
    std::cout << m_type.c_str() << " stamina was adjusted\n";
}

/***********************************
 * AdjustHealth
 *
 * Increments/Decrements health by 
 * change amount.
 *
 * changeAmt - the amount to increment
 *           or decrement health.
 ***********************************/
void BaseMole::AdjustHealth(int changeAmt)
{
    //m_stamina += changeAmt;
    std::cout << m_type.c_str() << " health was adjusted\n";
}

File: striker_library_projects/striker_library_projects/mole_library/Moles/BaseMole.cs
//-----–----------–------------------
// ImperfectlyCoded © 2018
//
// BaseMole.cs
//-----–----------–------------------
namespace mole_library.Moles
{
    using System;
    using System.Collections.Generic;
    using Interfaces;
    using Mounds;
    using States;

    /// <summary>
    /// Base implementation that all moles will inherit.
    /// </summary>
    public abstract class BaseMole
    {
        #region Fields

        /// <summary>
        /// The mounds the mole knows about in the level.
        /// </summary>
        protected List<Mound> m_mounds;

        #endregion

        #region Public Properties

        /// <summary>
        /// The mole's health
        /// </summary>
        public int Health { get; protected set; } = 20;

        /// <summary>
        /// The mole's stamina
        /// </summary>
        public int Stamina { get; protected set; } = 20;

        /// <summary>
        /// The mole's current state
        /// </summary>
        public BaseState CurrentState { get; protected set; } = new MoveOutMoundState();

        /// <summary>
        /// The mound that the mole currently occupies
        /// </summary>
        public Mound CurrentMound
        {
            get;
            protected set;
        }

        public List<Mound> Mounds
        {
            get
            {
                return m_mounds;
            }
        }

        #endregion

        #region Constructors

        /// <summary>
        /// Base constructor
        /// </summary>
        public BaseMole()
        {

        }

        /// <summary>
        /// Constructor to set the health, stamina and the mound
        /// for the mole
        /// </summary>
        /// <param name="health">The amount of health</param>
        /// <param name="stamina">The amount of stamina</param>
        /// <param name="mound">The mound for this mole to occupy</param>
        public BaseMole(int health, int stamina, Mound mound)
        {
            Health = health;
            Stamina = stamina;
            CurrentMound = mound;
        }

        #endregion

        #region Abstract Methods/Properties

        /// <summary>
        /// The type of mole
        /// </summary>
        public abstract string Type { get; protected set; }

        /// <summary>
        /// The job that defines the type of mole
        /// and that mole's attributes and abilities
        /// </summary>
        public abstract IJob Job { get; protected set; }

        /// <summary>
        /// Mole throws a projectile at the player.
        /// </summary>
        public abstract void ThrowProjectile();

        #endregion

        #region Public Methods

        /// <summary>
        /// Initialize the mole with any attributes necessary.
        /// </summary>
        /// <param name="mounds">The mounds in the level for the mole to know about.</param>
        public void Initialize(List<Mound> mounds)
        {
            if (mounds.Count > 0)
            {
                m_mounds = mounds;
            }
        }

        /// <summary>
        /// Resets all of the mole's timers
        /// </summary>
        public void ResetAllTimers()
        {

        }

        /// <summary>
        /// The action to perform based on the current state.
        /// </summary>
        public void Action()
        {
            CurrentState.Action(this);
        }

        /// <summary>
        /// Changes the mole's job.
        /// </summary>
        /// <param name="job">The new job.</param>
        public void ChangeJobs(IJob job)
        {
            Console.WriteLine("Old job: " + Job.Name);
            Console.WriteLine("New job: " + job.Name);
            Job = job;
        }

        /// <summary>
        /// Changes the mole's state
        /// </summary>
        /// <param name="state">The new state.</param>
        public void SetState(BaseState state)
        {
            Console.WriteLine("Exiting old state: " + CurrentState.Name);
            CurrentState.Exit(this);

            Console.WriteLine("Entering new state: " + state.Name);
            CurrentState = state;
            CurrentState.Enter(this);
        }

        /// <summary>
        /// Moves the mole inside of the current mound.
        /// </summary>
        public void MoveInMound()
        {
            Job.MoveInMound();
        }

        /// <summary>
        /// Moves the mole out of the current mound.
        /// </summary>
        public void MoveOutMound()
        {
            Job.MoveOutMound();
        }

        /// <summary>
        /// Moves mole to an unoccupied mound.
        /// </summary>
        public void MoveToDifferentMound()
        {
            Job.MoveToDifferentMound();
        }

        /// <summary>
        /// Mole performs a counter attack
        /// </summary>
        public void Counter()
        {
            Job.Counter();
        }

        /// <summary>
        /// Mole charges up attack, then attacks.
        /// </summary>
        public void Charge()
        {
            Job.Charge();
        }

        /// <summary>
        /// Mole evades oncoming attack.
        /// </summary>
        public void Evade()
        {
            Job.Evade();
        }

        /// <summary>
        /// Adjusts mole's stamina.
        /// </summary>
        /// <param name="changeAmt"></param>
        public void AdjustStamina(int changeAmt)
        {
            Stamina += changeAmt;
        }

        /// <summary>
        /// Adjust mole's health.
        /// </summary>
        /// <param name="changeAmt"></param>
        public void AdjustHealth(int changeAmt)
        {
            Health += changeAmt;
        }

        #endregion
    }
}
